gensim/models/keyedvectors.py

```diff
@@ -1130,16 +1256,22 @@ def evaluate_word_analogies(self, analogies, restrict_vocab=300000, case_insensi
                 self.vocab = ok_vocab
                 ignore = {a, b, c}  # input words to be ignored
                 predicted = None
-                # find the most likely prediction using 3CosAdd (vector offset) method
-                # TODO: implement 3CosMul and set-based methods for solving analogies
-                sims = self.most_similar(positive=[b, c], negative=[a], topn=5, restrict_vocab=restrict_vocab)
+                # find the most likely prediction using 3CosAdd (vector offset) method or 3CosMul
+
+                if method == '3CosAdd':
+                    sims = self.most_similar(positive=[b, c], negative=[a], topn=topk, restrict_vocab=restrict_vocab)    
+                if method == '3CosMul':
+                    sims = self.most_similar_cosmul(positive=[b, c], negative=[a], topn=topk, restrict_vocab=restrict_vocab)                    
                 self.vocab = original_vocab
-                for element in sims:
+
+                for element in sims:                  
                     predicted = element[0].upper() if case_insensitive else element[0]
                     if predicted in ok_vocab and predicted not in ignore:
                         if predicted != expected:
                             logger.debug("%s: expected %s, predicted %s", line.strip(), expected, predicted)
-                        break
+                        if predicted == expected:
```

Comment:
Why this change?

Reply:
break command changed because we want to consider also the top-5 similar words for each case.

Reply:
Sorry, I don't follow. What option does the fall-through-without-break code path solve?

There are too many levels of indentation here, too complicated. Can you factor some of this out into a named function?

Reply:
As it is implemented until now, when there is one of word among the top-k words similar words that belongs to the ok_vocab and not in "ignore" list, then we have a break command and we exit the for-loop. The latter means that we consider only this word and not all the top-k words in the comparison with the expected.

According to the change i ve made, we check if there is a match between the expected word and any of the k words from the top-k similar words list, which is different and in my opinion it is closer to the state of the art of the analogy questions used for the evaluation. I am looking forward to you comment. Thanks a lot!